tests/unit/local/docker/test_container_client.py

```diff
@@ -20,9 +20,185 @@
 )
 from samcli.local.docker.exceptions import (
     ContainerArchiveImageLoadFailedException,
+    ContainerInvalidSocketPathException,
 )
 
 
+class BaseContainerClientTestCase(TestCase):
+    """Base test case with common helper methods for container client testing"""
+
+    def setUp(self):
+        """Set up common test fixtures"""
+        self.finch_socket = "unix:///tmp/finch.sock"
+        self.docker_socket = "unix:///var/run/docker.sock"
+        self.default_version = "1.35"
+
+    def create_mock_api_client(self, base_url=None):
+        """Create a mock API client with standard configuration"""
+        mock_api_instance = Mock()
+        mock_api_instance.base_url = base_url or self.docker_socket
+        return mock_api_instance
+
+    def create_docker_client_spy(
+        self, env_vars=None, expected_base_url=None, client_class=DockerContainerClient, finch_socket_path=None
+    ):
+        """
+        Create a container client with constructor spy for testing initialization.
+
+        Args:
+            env_vars: Environment variables to set (default: empty)
+            expected_base_url: Expected base_url in constructor call
+            client_class: Container client class to instantiate
+            finch_socket_path: Custom finch socket path (default: self.finch_socket)
+
+        Returns:
+            tuple: (client, spy_init) where spy_init contains constructor call info
+        """
+        env_vars = env_vars or {}
+        finch_socket_path = finch_socket_path or self.finch_socket
+
+        with patch("samcli.local.docker.container_client.get_finch_socket_path", return_value=finch_socket_path), patch(
+            "os.environ", env_vars
+        ), patch("docker.api.client.APIClient", return_value=self.create_mock_api_client(expected_base_url)):
+
+            with patch.object(
+                docker.DockerClient, "__init__", side_effect=docker.DockerClient.__init__, autospec=True
+            ) as spy_init:
+                client = client_class()
```

Comment:
Do we understand this piece of code? The nested double `with`s, and returning `spy_init`, which is created inside the `with` all looks questionable.